File: cases/taint/scala/recursive-payload-transform-negative/RecursivePayloadTransformNegative.scala
package dataflowbench

object RecursivePayloadTransformNegative {
  def dfb_source(): Int = { // DFB-SOURCE: recursive-payload-transform-input
    7
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: recursive-payload-transform-sink

  def walk(value: Int, depth: Int): Int = {
    if (depth == 0) {
      var clean = value
      clean = 0 // DFB-KILL: recursive-payload-transform-base-clean
      clean // DFB-WITNESS: recursive-payload-transform-base
    } else {
      val next = value + 1
      val recursive = walk(next, depth - 1) // DFB-WITNESS: recursive-payload-transform-transfer
      recursive + 1 // DFB-WITNESS: recursive-payload-transform-compose
    }
  }

  def run(): Unit = {
    dfb_sink(walk(dfb_source(), 3))
  }
}

File: cases/taint/c/recursive-payload-transform-negative/recursive_payload_transform_negative.c
int dfb_source(void) { // DFB-SOURCE: recursive-payload-transform-input
    return 7;
}

void dfb_sink(int value) { (void)value; } // DFB-SINK: recursive-payload-transform-sink

static int walk(int value, int depth) {
    if (depth == 0) {
        value = 0; // DFB-KILL: recursive-payload-transform-base-clean
        return value; // DFB-WITNESS: recursive-payload-transform-base
    }
    int next = value + 1;
    int recursive = walk(next, depth - 1); // DFB-WITNESS: recursive-payload-transform-transfer
    return recursive + 1; // DFB-WITNESS: recursive-payload-transform-compose
}

void run(void) {
    dfb_sink(walk(dfb_source(), 3));
}
